src/no_std_error.rs

```diff
@@ -0,0 +1,199 @@
+use core::fmt::{Debug, Display};
+
+/// `Error` is a trait representing the basic expectations for error values,
+/// i.e., values of type `E` in `Result<T, E>`. Errors must describe
+/// themselves through the `Display` and `Debug` traits, and may provide
+/// cause chain information:
+///
+/// The [`source`] method is generally used when errors cross "abstraction
+/// boundaries". If one module must report an error that is caused by an error
+/// from a lower-level module, it can allow access to that error via the
+/// [`source`] method. This makes it possible for the high-level module to
+/// provide its own errors while also revealing some of the implementation for
+/// debugging via [`source`] chains.
+///
+/// [`source`]: trait.Error.html#method.source
+pub trait Error: Debug + Display {
```

Comment:
I'm very hesitant to copy the standard libraries `Error` like this, especially the implementations on arbitrary types. One thing I've been negative about is how failure created a new trait and then told everyone that they needed to switch to it to be compatible. This felt like it was a bad "team player" in the broader ecosystem.

I suppose we'd quickly run into coherence problems if we make this a marker trait and provided a blanket implementation?

```rust
pub trait Error: Debug + Display {}
impl<T: Debug + Display> Error for T {}
```

Another avenue is to completely drop `Error` and just implement `Debug + Display`; any idea what that world might look like?

Reply:
I've considered that when I initially tried implementing this, but you quickly lose out on most of snafu's feature such as the ResultExt and OptionExt traits. We'd at least need some trait that we implement in addition to Display and Debug so we can somehow make ResultExt and OptionExt work too. Also not having source chaining is probably pretty bad too. So that's why in my initial prototyping it evolved from not having an Error trait at all to eventually just a Copy of the full thing (although I did consider dropping the deprecated methods).

Reply:
The Error trait introduced here is not really meant to be used anywhere really. In livesplit-core where I've been using this no-std snafu branch the Error trait never is publicly visible to anyone. It is mostly just a convenience trait to not lose out on any of the features of snafu.

Reply:
Actually the blanket implementation Error trait could maybe work. I'll see if it causes any problems.

Reply:
Although I don't think there even is any harm to having a copy of the Error trait in snafu. You can still treat core as if it didn't have an Error trait at all. But if you want to use an Error trait snafu optionally has one available for you to use. And if someone activates the std feature (either directly through a feature of your crate or indirectly), your code automatically uses the std Error trait and there is no "bad team playing" going on.

Reply:
Although one problem I can see is that errors from other core crates might not implement an Error trait at all on core, so you can't use ResultExt and co. regardless.

Reply:
It feels like we will need to release *something* and get feedback from people as to the usefulness. I think it's best we call the no-std support / exact implementation experimental.

For example, one thing that concerns me is what happens if one crate using no-std SNAFU exports an error type. Will other crates be able to make use of it, or will they need to wrap it in that newtype?

Reply:
I'm not sure I understand what you mean with that. Why would they need to wrap it in a newtype? If the error type is from snafu in a no_std context, then there simply is no Error trait, just like usual. Or are you saying what the story is with two crates using snafu in a no_std context? That should work seamlessly too, unless they are using different snafu versions.